ucp/core.py

```diff
@@ -36,35 +35,57 @@ def _get_ctx():
 
 
 async def exchange_peer_info(
-    endpoint, msg_tag, ctrl_tag, guarantee_msg_order, listener
+    endpoint, msg_tag, ctrl_tag, guarantee_msg_order, port, listener
 ):
     """Help function that exchange endpoint information"""
 
-    msg_tag = int(msg_tag)
-    ctrl_tag = int(ctrl_tag)
-    guarantee_msg_order = bool(guarantee_msg_order)
-    my_info = struct.pack("QQ?", msg_tag, ctrl_tag, guarantee_msg_order)
+    # Pack peer information incl. a checksum
+    fmt = "QQ?QQ"
+    my_info = struct.pack(
+        fmt,
+        msg_tag,
+        ctrl_tag,
+        guarantee_msg_order,
+        port,
+        hash64bits(msg_tag, ctrl_tag, guarantee_msg_order, port),
+    )
     peer_info = bytearray(len(my_info))
 
+    # Send/recv peer information. Notice, we force an `await` between the two
+    # streaming calls (see <https://github.com/rapidsai/ucx-py/pull/509>)
     if listener is True:
         await ucx_api.stream_send(endpoint, my_info, len(my_info))
         await ucx_api.stream_recv(endpoint, peer_info, len(peer_info))
     else:
         await ucx_api.stream_recv(endpoint, peer_info, len(peer_info))
         await ucx_api.stream_send(endpoint, my_info, len(my_info))
 
-    peer_msg_tag, peer_ctrl_tag, peer_guarantee_msg_order = struct.unpack(
-        "QQ?", peer_info
+    # Unpacking and sanity check of the peer information
+    ret = {}
+    (
+        ret["msg_tag"],
+        ret["ctrl_tag"],
+        ret["guarantee_msg_order"],
+        ret["port"],
+        ret["checksum"],
+    ) = struct.unpack(fmt, peer_info)
+
+    expected_checksum = hash64bits(
+        ret["msg_tag"], ret["ctrl_tag"], ret["guarantee_msg_order"], ret["port"]
     )
 
-    if peer_guarantee_msg_order != guarantee_msg_order:
+    if expected_checksum != ret["checksum"]:
+        raise RuntimeError(
+            f'Checksum invalid! {hex(expected_checksum)} != {hex(ret["checksum"])}'
+        )
+
+    if port != ret["port"]:
+        raise RuntimeError(f'Port mismatch! {port} != {ret["port"]}')
+
+    if ret["guarantee_msg_order"] != guarantee_msg_order:
         raise ValueError("Both peers must set guarantee_msg_order identically")
```

Comment:
I don't think either of those would ever be raised, given the checksum above will raise an exception before. I suggest removing them.

Reply:
Notice, the checksum is for the message itself thus both port and guarantee_msg_order can disagree between the two endpoints, which is also tests: https://github.com/rapidsai/ucx-py/blob/f38645426e48e0f9f4731b6bd1bf8e84b097c169/tests/test_guarantee_msg_order.py#L8

Reply:
Oh yes, you're right about `guarantee_msg_order`, my mistake. I think the port case will never happen though as that would require a client to connect to the wrong listener somehow _and_ be lucky that there's a UCX listener on that port exactly, but that's another small check so that's probably fine to stay as well.

Reply:
>  I think the port case will never happen though as that would require a client to connect to the wrong listener somehow

Agree, I added it while debugging and thought it was useful to have as an extra sanity check :)